src/Analysis/Engine/Impl/Infrastructure/Extensions/ArrayExtensions.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+
+namespace Microsoft.PythonTools.Analysis.Infrastructure {
+    internal static class ArrayExtensions {
+        public static T[] ImmutableAdd<T>(this T[] array, in T item) {
+            if (array.Length == 0) {
+                return new[] {item};
+            }
+
+            var copy = new T[array.Length + 1];
+            Array.Copy(array, copy, array.Length);
```

Comment:
Out of curiosity, is copying cheaper than List?

Reply:
In terms of memory? No, `List<T>` is cheaper, but it isn't immutable. Benchmark that created a 5-level tree with 5 children for each node (15626 nodes total) by adding each node one by one produced about 130 kilobytes of garbage.